src/NHibernate/Async/Type/NullableType.cs

```diff
@@ -0,0 +1,152 @@
+		/// /> 
+		/// <remarks>
+		/// This has been sealed because no other class should override it.  This 
+		/// method calls <see cref="NullSafeGetAsync(DbDataReader, String,CancellationToken)" /> for a single value.  
+		/// It only takes the first name from the string[] names parameter - that is a 
+		/// safe thing to do because a Nullable Type only has one field.
+		/// </remarks>
+		public override sealed Task<object> NullSafeGetAsync(DbDataReader rs, string[] names, ISessionImplementor session, object owner, CancellationToken cancellationToken)
+		{
+			if (cancellationToken.IsCancellationRequested)
+			{
+				return Task.FromCanceled<object>(cancellationToken);
+			}
+			return NullSafeGetAsync(rs, names[0], cancellationToken);
+		}
+
+		/// <summary>
+		/// Extracts the values of the fields from the DataReader
+		/// </summary>
+		/// <param name="rs">The DataReader positioned on the correct record</param>
+		/// <param name="names">An array of field names.</param>
+		/// <param name="cancellationToken">A cancellation token that can be used to cancel the work</param>
+		/// <returns>The value off the field from the DataReader</returns>
+		/// <remarks>
+		/// In this class this just ends up passing the first name to the NullSafeGet method
+		/// that takes a string, not a string[].
+		/// 
+		/// I don't know why this method is in here - it doesn't look like anybody that inherits
+		/// from NullableType overrides this...
+		/// 
+		/// TODO: determine if this is needed
+		/// </remarks>
+		public virtual Task<object> NullSafeGetAsync(DbDataReader rs, string[] names, CancellationToken cancellationToken)
+		{
+			if (cancellationToken.IsCancellationRequested)
+			{
+				return Task.FromCanceled<object>(cancellationToken);
+			}
+			return NullSafeGetAsync(rs, names[0], cancellationToken);
+		}
+
+		/// <summary>
+		/// Gets the value of the field from the <see cref="DbDataReader" />.
+		/// </summary>
+		/// <param name="rs">The <see cref="DbDataReader" /> positioned on the correct record.</param>
+		/// <param name="name">The name of the field to get the value from.</param>
+		/// <param name="cancellationToken">A cancellation token that can be used to cancel the work</param>
+		/// <returns>The value of the field.</returns>
+		/// <remarks>
+		/// <para>
+		/// This method checks to see if value is null, if it is then the null is returned
+		/// from this method.
+		/// </para>
+		/// <para>
+		/// If the value is not null, then the method <see cref="Get(DbDataReader, Int32)"/> 
+		/// is called and that method is responsible for retrieving the value.
+		/// </para>
+		/// </remarks>
+		public virtual async Task<object> NullSafeGetAsync(DbDataReader rs, string name, CancellationToken cancellationToken)
+		{
+			cancellationToken.ThrowIfCancellationRequested();
+			int index = rs.GetOrdinal(name);
+
+			if (await (rs.IsDBNullAsync(index, cancellationToken)).ConfigureAwait(false))
```

Comment:
Is it worth it to have async only for the `null` check, while not handling the get?

I think we should either not have those async get/set, or handle the get too. Probably by using `GetFieldValue<T>`, which has an async method, instead of hard typed variants (`GetInt32`, ...) in the synchronous NHibernate.